src/main/java/com/openlattice/organizations/HazelcastOrganizationService.java

```diff
@@ -320,25 +323,36 @@ private void addMembers( AclKey orgAclKey, Set<Principal> members ) {
     }
 
     public void removeMembers( UUID organizationId, Set<Principal> members ) {
-        removeRolesFromMembers(
-                getRolesInFull( organizationId ).stream().map( Role::getAclKey ),
-                members
-                        .stream()
-                        .filter( m -> m.getType().equals( PrincipalType.USER ) )
-                        .map( securePrincipalsManager::lookup ) );
-        membersOf.executeOnKey( organizationId, new OrganizationMemberRemover( members ) );
+        final var users = members.stream()
+                .filter( m -> m.getType().equals( PrincipalType.USER ) )
+                .collect( Collectors.toList() );
+        final var securablePrincipals = securePrincipalsManager.getSecurablePrincipals( users );
 
+        removeRolesFromMembers( getRolesInFull( organizationId ).stream().map( Role::getAclKey ), securablePrincipals );
+        membersOf.executeOnKey( organizationId, new OrganizationMemberRemover( members ) );
         final AclKey orgAclKey = new AclKey( organizationId );
-        members.stream().filter( PrincipalType.USER::equals )
-                .map( securePrincipalsManager::lookup )
-                .forEach( target -> securePrincipalsManager.removePrincipalFromPrincipal( orgAclKey, target ) );
+        removeOrganizationFromMembers( orgAclKey, securablePrincipals );
+
+        eventBus.post(
+                new MembersRemovedFromOrganizationEvent(
+                        organizationId,
+                        new SecurablePrincipalList( securablePrincipals )
+                )
+        );
+    }
 
-        eventBus.post( new MembersRemovedFromOrganizationEvent( organizationId, new PrincipalSet( members ) ) );
+    private void removeRolesFromMembers( Stream<AclKey> roles, Collection<SecurablePrincipal> members ) {
+        members.forEach( securablePrincipal ->
+                roles.forEach( role ->
+                        securePrincipalsManager.removePrincipalFromPrincipal( role, securablePrincipal.getAclKey() )
```

Comment:
We should probably allow for a set of AclKeys to be passed to securePrincipalsManager so that we can bulk the calls to hazelcast instead of making individual calls for each principal/role pair.

Also, removePrincipalFromPrincipal ends up constructing a singleton set for each securable principal which it passes to an entry processor, so we should probably just support passing the collection through directly instead of iterating and calling individually.

Reply:
Gotcha, will do!